File: haskell-server/src/Bolour/Util/CompactStringSet.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}


-- Experiment to try and reduce memory overhead of creating sets.
-- Obsolete. The experiment failed!
-- If necessary, try using mutable arrays IOArray instead.
-- But the main overhead was reading a file as String vs as lazy byte string.
module Bolour.Util.CompactStringSet (
    CompactStringSet(..)
  , mkFromList
  , mkCompactStringSet
  , insert
  , insertAll
  , contains
  ) where

import Data.Bits
import qualified Data.Map.Strict as Map
import qualified Data.List as List
import Data.Hashable
import qualified Data.List.Split as Split

import qualified Bolour.Util.MiscUtil as Util

-- | Space efficient implementation of set of small strings.
--   Packs a number of strings into a single bucket.
data CompactStringSet = CompactStringSet {
  buckets :: Map.Map Int String
}

mkCompactStringSet :: CompactStringSet
mkCompactStringSet = CompactStringSet Map.empty

-- TODO. Size of hash table should be an input parameter to mkCompactStringSet.

hashMask :: Int
-- hashMask = 0x8FFFF -- 589,823
hashMask = 0xFFFFFF -- 16,777,215

hashIt :: String -> Int
hashIt elem = hash elem .&. hashMask

delimiterChar :: Char
delimiterChar = '\0'

delimiter :: String
delimiter = [delimiterChar]

pack :: String -> String -> String
pack packedElems elem = packedElems ++ delimiter ++ elem

packAll :: [String] -> String
packAll strings = List.intercalate delimiter strings

unpack :: String -> [String]
unpack = Split.splitOn delimiter

insert :: CompactStringSet -> String -> CompactStringSet
insert CompactStringSet { buckets } elem =
  let index = hashIt elem
      bucket = Map.lookup index buckets
      bucketPlus = case bucket of
        Nothing -> elem
        Just packedElems -> pack packedElems elem
      bucketsPlus = Map.insert index bucketPlus buckets
  in CompactStringSet bucketsPlus

contains :: CompactStringSet -> String -> Bool
contains CompactStringSet { buckets } elem =
    let index = hashIt elem
        bucket = Map.lookup index buckets
    in case bucket of
       Nothing -> False
       Just packedElems -> elem `List.elem` unpack packedElems

-- TODO. Insertion of bulk data uses too much memory.
-- For now unusable for masked words. To be diagnosed.
insertAll :: CompactStringSet -> [String] -> CompactStringSet
insertAll = List.foldl' insert

mkFromList :: [String] -> CompactStringSet
mkFromList values =
  let mapToList = Util.mapFromValueList hashIt values
      theMap = packAll <$> mapToList
  in CompactStringSet theMap

File: scala-server/app/com/bolour/util/scala/common/CompactStringSet.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */

package com.bolour.util.scala.common

import scala.collection.mutable

/**
  * Space-efficient implementation of a set of small strings.
  * Hashes strings into compactly packed buckets.
  * Not time-efficient for inserting collections. Needs to be parallelized.
  * A failed attempt at reducing the memory footprint of dictionary masked words.
  */
class CompactStringSet {

  import CompactStringSet._

  val buckets = mutable.Map[Int, String]()

  def insert(elem: String): Unit = {
    val index = hash(elem)
    val bucket = buckets.get(index)
    val bucketPlus = bucket match {
      case None => elem
      case Some(elems) => pack(elems, elem)
    }
    buckets(index) = bucketPlus
  }

  def contains(elem: String): Boolean = {
    val index = hash(elem)
    val bucket = buckets.get(index)
    bucket match {
      case None => false
      case Some(elems) => unpack(elems).contains(elem)
    }
  }

  def ++(elems: Iterator[String]): Unit = elems.foreach {elem => insert(elem)}
  def ++(elems: java.util.stream.Stream[String]): Unit = elems.forEach {elem => insert(elem)}

  private def hash(string: String): Int = string.hashCode & hashMask

  override def toString = buckets.toString

}

object CompactStringSet {
  // TODO. Hash bits and delimiter should be instance parameters.
  val hashMask = 0x8FFFF  // 589823
  val delimiterChar = '\u0000'
  val delimiter = delimiterChar.toString

  def pack(packedStrings: String, elem: String): String =
    packedStrings ++ delimiter ++ elem

  def unpack(packedStrings: String): List[String] = packedStrings.split(delimiter).toList

  def apply(iterator: Iterator[String]): CompactStringSet = {
    val set = new CompactStringSet()
    set ++ iterator
    set
  }

  def apply(stream: java.util.stream.Stream[String]): CompactStringSet = {
    val set = new CompactStringSet()
    set ++ stream
    set
  }

  def apply(list: List[String]): CompactStringSet = CompactStringSet(list.iterator)
}
